README.md
tomdoc.sh
=========

tomdoc.sh will parse [TomDoc]'d shell scripts and generate pretty documentation
from it.

TomDoc for shell? - Absolutely. Even though TomDoc was originally specified to
document Ruby code, it turns out to be a good fit for shell scripts too. Proof:

```sh
# Public: Current API version in format "x.y.z".
export API_VERSION="1.2.3"

# Public: Execute commands in debug mode.
#
# Takes a single argument and evaluates it only when the test script is started
# with --debug. This is primarily meant for use during the development of test
# scripts.
#
# $1 - Commands to be executed.
#
# Examples
#
#   test_debug "cat some_log_file"
#
# Returns the exit code of the last command executed in debug mode or 0
#   otherwise.
test_debug() {
	test "$debug" = "" || eval "$1"
}
```

Passing the above to tomdoc.sh results in:

```
API_VERSION
-----------

Public: Current API version in format "x.y.z".

test_debug()
------------

Public: Execute commands in debug mode.

Takes a single argument and evaluates it only when the test script is started with --debug. This is primarily meant for use during the development of test scripts.

* `$1` - Commands to be executed.

Examples

    test_debug "cat some_log_file"

Returns the exit code of the last command executed in debug mode or 0 otherwise.
```

For maximum portability, tomdoc.sh was written in POSIX shell and only depends
on ubiquitous Unix tools like `sed(1)` and `grep(1)`.

The project was inspired by [tomdoc.rb].


Installation
------------

Installing tomdoc.sh is simple:

    $ git clone git://github.com/tests-always-included/tomdoc.sh.git
    $ cd tomdoc.sh
    $ make install

That will install tomdoc.sh to `$HOME/bin` by default. For a different target
directory, set `bindir` accordingly, e.g.

    # make install bindir=/usr/local/bin


Usage
-----

    Usage: tomdoc.sh [options] [--] [<shell-script>...]

        -h, --help               show help text
        --version                show version
        -t, --text               produce plain text (default format)
        -m, --markdown           produce markdown
        -a, --access <level>     filter by access level

    Parse TomDoc'd shell scripts and generate pretty documentation from it.

For each TomDoc'd shell script you pass to tomdoc.sh, it will produce pretty
documentation in plain text (option `--text`, the default) or markdown format
(`--markdown`), writing the results to the standard output.

With `--access`, output can be limited to shell functions and variables
documented as "Public", "Internal", or "Deprecated".

The `<shell-script>` operands are processed in command-line order. If
`<shell-script>` is a single dash (-) or absent, tomdoc.sh reads from the
standard input.

For testing, you can run tomdoc.sh on itself:

    $ ./tomdoc.sh --text tomdoc.sh
    $ ./tomdoc.sh --markdown tomdoc.sh

The generated output can be seen [here][fixtures].


Tests
-----

The `test` folder contains some automated test scripts powered by [Sharness].

You can run the tests this way:

    $ make test


License
-------

* tomdoc.sh is licensed under the terms of the MIT License. See [LICENSE] file.
* [Sharness] and all tests are licensed under the terms of the GNU General
  Public License version 2 or higher. See file [COPYING] for full license text.


Contact
-------

* Web: <https://github.com/tests-always-included/tomdoc.sh>
* Mail: <fidian@rumkin.com>


[COPYING]: ./test/COPYING
[LICENSE]: ./LICENSE
[Sharness]: https://github.com/mlafeldt/Sharness
[TomDoc]: http://tomdoc.org
[fixtures]: ./test/fixtures
[tomdoc.rb]: https://github.com/defunkt/tomdoc


HISTORY.md
History
=======

0.1.10 (2022-03-29)
-------------------

* Tracked changes in HISTORY.
*

0.1.9 (2021-11-15)
------------------

* Ignore `# shellcheck` directives.
* Added support for `declare` and `typeset`, then fixed patterns for Mac.
* Reformatted code, added lint checks, added config files for tooling to help with development.
* Added notes to watch out for issues when using Mac's built-in `sed`.

0.1.8 (2018-03-12)
------------------

* Bash functions may use most character, so the pattern to match function names
  was updated. See issue #8.
* Variable names are more restrictive than function names, so a separate
  pattern was made to match those instead of reusing the function name pattern.
* BSD `sed` will always add a newline, so add workarounds to remove that
  newline.
* Removed `safe_echo` and `safe_echo_n` and use `printf` instead.

0.1.7 (2016-08-22)
------------------

* Fixed a problem that happens when lines contain escape codes that the POSIX
  version of `echo` honors.

0.1.6 (2016-08-04)
------------------

* Transferred to the `tests-always-included` organization on GitHub.
* Improved the generated markdown

0.1.5 (2013-11-23)
------------------

* Update to Sharness v0.3.0.
* Let Travis only test master branch and pull requests.

0.1.4 (2012-04-08)
------------------

* Support TomDoc'd `export parameter` without assignment.
* Support TomDoc'd `: ${parameter:=value}` and `: ${parameter=value}`.
* Add dedicated parsing tests.

0.1.3 (2012-03-15)
------------------

* Allow to filter TomDoc'd functions and variables by access level (via
  `--access`). For example, with "--access Public" only TomDoc strings starting
  with "Public:" are processed.
* Let Travis CI use Erlang workers and run `prove(1)`.
* Update to Sharness v0.2.1.

0.1.2 (2012-02-27)
------------------

* Support TomDoc'd shell variables.
* Use `sed(1)` in a portable way.
* Allow to read from multiple files and standard input.
* Add Makefile to install or uninstall tomdoc.sh.
* Add Travis CI config.

0.1.1 (2012-02-16)
------------------

* Improve TomDoc and function name parsing. Support Bash's `function` keyword.
* Simplify doc generation.
* Add options `--help` and `--version`.
* Set shell option -e. Set -x when `TOMDOCSH_DEBUG` is defined in environment.
* Save test results to different files for debugging.

0.1.0 (2012-02-09)
------------------

* First version. Allows to generate documentation for TomDoc'd shell functions
  in (ugly) plain text or (ugly) markdown.


test/t0004-access.sh
#!/bin/sh

test_description="Match access level"

. ./sharness.sh

cat >script.sh <<EOF
# Some variable w/o access level.
VAR=0

# Public: Some public variable.
PUBLIC_VAR=1

# Internal: Some internal variable.
INTERNAL_VAR=2

# Deprecated: Some deprecated variable.
DEPRECATED_VAR=3
EOF

cat >expect <<EOF
--------------------------------------------------------------------------------
VAR

Some variable w/o access level.

--------------------------------------------------------------------------------
PUBLIC_VAR

Public: Some public variable.

--------------------------------------------------------------------------------
INTERNAL_VAR

Internal: Some internal variable.

--------------------------------------------------------------------------------
DEPRECATED_VAR

Deprecated: Some deprecated variable.

EOF

test_expect_success "Match everything" "
    tomdoc.sh --text script.sh >result &&
    test_cmp expect result
"

cat >expect <<EOF
--------------------------------------------------------------------------------
PUBLIC_VAR

Public: Some public variable.

EOF

test_expect_success "Match Public" "
    tomdoc.sh --text --access Public script.sh >result &&
    test_cmp expect result
"

cat >expect <<EOF
--------------------------------------------------------------------------------
INTERNAL_VAR

Internal: Some internal variable.

EOF

test_expect_success "Match Internal" "
    tomdoc.sh --text --access Internal script.sh >result &&
    test_cmp expect result
"

cat >expect <<EOF
--------------------------------------------------------------------------------
DEPRECATED_VAR

Deprecated: Some deprecated variable.

EOF

test_expect_success "Match Deprecated" "
    tomdoc.sh --text --access Deprecated script.sh >result &&
    test_cmp expect result
"

test_done


test/t0001-generate.sh
#!/bin/sh

test_description="Generate documentation"

. ./sharness.sh

FIXTURES="$TEST_DIRECTORY/fixtures"

test_expect_success "Generate plain text documentation" "
    tomdoc.sh --text '$BUILD_DIR/tomdoc.sh' >result.txt &&
    test_cmp '$FIXTURES/tomdoc.sh.txt' result.txt
"

test_expect_success "Generate markdown documentation" "
    tomdoc.sh --markdown '$BUILD_DIR/tomdoc.sh' >result.md &&
    test_cmp '$FIXTURES/tomdoc.sh.md' result.md
"

test_done


test/t0002-functions.sh
#!/bin/sh

test_description="Parse shell functions"

. ./sharness.sh

cat >expect <<EOF
--------------------------------------------------------------------------------
func()

Some function.

EOF

parse() {
    cat >script <<EOF
# Some function.
$1
EOF

    test_expect_success "Parse '$1'" "
        tomdoc.sh script >result &&
        test_cmp expect result
    "
}

parse 'func()'
parse 'func() {}'
parse 'function func()'
parse '	function	func	()	'

test_done


tomdoc.sh
#!/bin/sh
#/ Usage: tomdoc.sh [options] [--] [<shell-script>...]
#/
#/     -h, --help               show help text
#/     --version                show version
#/     -t, --text               produce plain text (default format)
#/     -m, --markdown           produce markdown
#/     -a, --access <level>     filter by access level
#/
#/ Parse TomDoc'd shell scripts and generate pretty documentation from it.
#
# Written by Mathias Lafeldt <mathias.lafeldt@gmail.com>, later project was
# transfered to Tyler Akins <fidian@rumkin.com>.

set -e
test -n "$TOMDOCSH_DEBUG" && set -x

# Current version of tomdoc.sh.
TOMDOCSH_VERSION="0.1.10"

generate=generate_text
access=

while test "$#" -ne 0; do
    case "$1" in
    -h | --h | --he | --hel | --help)
        grep '^#/' <"$0" | cut -c4-
        exit 0
        ;;
    --version)
        printf "tomdoc.sh version %s\n" "$TOMDOCSH_VERSION"
        exit 0
        ;;
    -t | --t | --te | --tex | --text)
        generate=generate_text
        shift
        ;;
    -m | --m | --ma | --mar | --mark | --markd | --markdo | --markdow | --markdown)
        generate=generate_markdown
        shift
        ;;
    -a | --a | --ac | --acc | --acce | --acces | --access)
        test "$#" -ge 2 || {
            printf "error: %s requires an argument\n" "$1" >&2
            exit 1
        }
        access="$2"
        shift 2
        ;;
    --)
        shift
        break
        ;;
    - | [!-]*)
        break
        ;;
    -*)
        printf "error: invalid option '%s'\n" "$1" >&2
        exit 1
        ;;
    esac
done

# Regular expression matching at least one whitespace.
SPACE_RE='[[:space:]][[:space:]]*'

# Regular expression matching optional whitespace.
OPTIONAL_SPACE_RE='[[:space:]]*'

# The inverse of the above, must match at least one character
NOT_SPACE_RE='[^[:space:]][^[:space:]]*'

# Regular expression matching shell function or variable name.  Functions may
# use nearly every character.  See [issue #8].  Disallowed characters (hex,
# octal, then a description or a character):
#
#   00 000 null       01 001 SOH        09 011 Tab        0a 012 Newline
#   20 040 Space      22 042 Quote      23 043 #          24 044 $
#   26 046 &          27 047 Apostrophe 28 050 (          29 051 )
#   2d 055 Hyphen     3b 073 ;          3c 074 <          3d 075 =
#   3e 076 >          5b 133 [          5c 134 Backslash  60 140 Backtick
#   7c 174 |          7f 177 Delete
#
# Exceptions allowed as leading character:  \x3d and \x5b
# Exceptions allowed as secondary character: \x23 and \x2d
#
# Must translate to raw characters because Mac OS X's sed does not work with
# escape sequences.  All escapes are handled by printf.
#
# Must use a hyphen first because otherwise it is an invalid range expression.
#
# [issue #8]: https://github.com/tests-always-included/tomdoc.sh/issues/8
FUNC_NAME_RE=$(printf "[^-\\001\\011 \"#$&'();<>\\134\\140|\\177][^\\001\\011 \"$&'();<=>[\\134\\140|\\177]*")

# Regular expression matching variable names.  Similar to FUNC_NAME_RE.
# Variables are far more restrictive.
#
# Leading characters can be A-Z, _, a-z.
# Secondary characters can be 0-9, =, A-Z, _, a-z
VAR_NAME_RE='[A-Z_a-z][0-9=A-Z_a-z]*'

# Strip leading whitespace and '#' from TomDoc strings.
#
# Can not use \?, use \{0,1\} instead to preserve Mac support.
#
# Returns nothing.
uncomment() {
    sed -e "s/^$OPTIONAL_SPACE_RE#[[:space:]]\{0,1\}//"
}

# Generate the documentation for a shell function or variable in plain text
# format and write it to stdout.
#
# $1 - Function or variable name
# $2 - TomDoc string
#
# Returns nothing.
generate_text() {
    cat <<EOF
--------------------------------------------------------------------------------
$1

$(printf "%s" "$2" | uncomment)

EOF
}

# Generate the documentation for a shell function or variable in markdown format
# and write it to stdout.
#
# $1 - Function or variable name
# $2 - TomDoc string
#
# Returns nothing.
generate_markdown() {
    printf "%s\n" '`'"$1"'`'
    printf "%s" " $1 " | tr -c - -
    printf "\n\n"

    last=""
    did_newline=false
    last_was_option=false

    printf "%s\n" "$2" | uncomment | sed -e "s/$SPACE_RE$//" | while IFS='' read -r line; do
        if printf "%s" "$line" | grep -q "^$OPTIONAL_SPACE_RE$NOT_SPACE_RE$SPACE_RE-$SPACE_RE"; then
            # This is for arguments
            if ! $did_newline; then
                printf "\n"
            fi

            if printf "%s" "$line" | grep -q "^$NOT_SPACE_RE"; then
                printf "%s" "* $line"
            else
                # Careful - BSD sed always adds a newline
                printf "    * "
                printf "%s" "$line" | sed "s/^$SPACE_RE//" | tr -d "\n"
            fi

            last_was_option=true

            # shellcheck disable=SC2030

            did_newline=false
        else
            case "$line" in
            "")
                # Check for end of paragraph / section
                if ! $did_newline; then
                    printf "\n"
                fi

                printf "\n"
                did_newline=true
                last_was_option=false
                ;;

            "  "*)
                # Examples and option continuation
                if $last_was_option; then
                    # Careful - BSD sed always adds a newline
                    printf "%s" "$line" | sed "s/^ */ /" | tr -d "\n"
                    did_newline=false
                else
                    printf "  %s\n" "$line"
                    did_newline=true
                fi
                ;;

            "* "*)
                # A list should not continue a previous paragraph.
                printf "%s\n" "$line"
                did_newline=true
                ;;

            *)
                # Paragraph text (does not start with a space)
                case "$last" in
                "")
                    # Start a new paragraph - no space at the beginning
                    printf "%s" "$line"
                    ;;

                *)
                    # Continue this line - include space at the beginning
                    printf " %s" "$line"
                    ;;
                esac

                did_newline=false
                last_was_option=false
                ;;
            esac
        fi

        last="$line"
    done

    # shellcheck disable=SC2031

    if ! $did_newline; then
        printf "\n"
    fi
}

# Read lines from stdin, look for shell function or variable definition, and
# print function or variable name if found; otherwise, print nothing.
#
# Can not use \?, use \{0,1\} instead to preserve Mac support.
# Can not use \(a\|b\), use two patterns instead for Mac support.
#
# Returns nothing.
parse_code() {
    sed -n \
        -e "s/^$OPTIONAL_SPACE_RE\(function$SPACE_RE\)\{0,1\}\($FUNC_NAME_RE\)$OPTIONAL_SPACE_RE().*$/\2()/p" \
        -e "s/^$OPTIONAL_SPACE_RE\(export$SPACE_RE\)\{0,1\}\($VAR_NAME_RE\)=.*$/\2/p" \
        -e "s/^${OPTIONAL_SPACE_RE}export$SPACE_RE\($VAR_NAME_RE\).*$/\1/p" \
        -e "s/^$OPTIONAL_SPACE_RE:$SPACE_RE\${\($VAR_NAME_RE\):\{0,1\}=.*$/\1/p" \
        -e "s/^${OPTIONAL_SPACE_RE}declare$SPACE_RE\(-[a-zA-Z]*$SPACE_RE\)\{0,1\}\($VAR_NAME_RE\)=.*$/\2/p" \
        -e "s/^${OPTIONAL_SPACE_RE}declare$SPACE_RE\(-[a-zA-Z]*$SPACE_RE\)\{0,1\}\($VAR_NAME_RE\).*$/\2/p" \
        -e "s/^${OPTIONAL_SPACE_RE}typeset$SPACE_RE\(-[a-zA-Z]*$SPACE_RE\)\{0,1\}\($VAR_NAME_RE\)=.*$/\2/p" \
        -e "s/^${OPTIONAL_SPACE_RE}typeset$SPACE_RE\(-[a-zA-Z]*$SPACE_RE\)\{0,1\}\($VAR_NAME_RE\).*$/\2/p"
}

# Read lines from stdin, look for TomDoc'd shell functions and variables, and
# pass them to a generator for formatting.
#
# Returns nothing.
parse_tomdoc() {
    doc=
    while read -r line; do
        case "$line" in
        '# shellcheck'*)
            continue
            ;;

        '#' | '# '*)
            doc="$doc$line
"
            ;;
        *)
            test -n "$line" && test -n "$doc" && {
                # Match access level if given.
                test -n "$access" &&
                    case "$doc" in
                    "# $access:"*) ;;

                    *)
                        doc=
                        continue
                        ;;
                    esac

                name="$(printf "%s" "$line" | parse_code)"
                test -n "$name" && "$generate" "$name" "$doc"
            }
            doc=
            ;;
        esac
    done
}

cat -- "$@" | parse_tomdoc

:
